File: RmlUi.Net/ElementDocument.cs
namespace ChickenWithLips.RmlUi;

public class ElementDocument : Element<ElementDocument>
{
    #region Properties

    #endregion

    #region Methods

    protected ElementDocument(IntPtr ptr, bool automaticallyRegisterInCache)
        : base(ptr, automaticallyRegisterInCache)
    {
    }

    /// <summary>
    /// Show the document.
    /// </summary>
    /// <param name="modalFlag">Flags controlling the modal state of the document, see the 'ModalFlag' description for details.</param>
    /// <param name="focusFlag">Flags controlling the focus, see the 'FocusFlag' description for details.</param>
    public void Show(ModalFlag modalFlag = ModalFlag.None, FocusFlag focusFlag = FocusFlag.Auto)
    {
        Native.ElementDocument.Show(NativePtr, modalFlag, focusFlag);
    }

    /// <summary>
    /// Hide the document.
    /// </summary>
    public void Hide()
    {
        Native.ElementDocument.Hide(NativePtr);
    }

    /// <summary>
    /// Close the document.
    /// </summary>
    /// <remarks>
    /// The destruction of the document is deferred until the next call to Context::Update().
    /// </remarks>
    public void Close()
    {
        Native.ElementDocument.Close(NativePtr);
    }

    internal static ElementDocument? Create(IntPtr ptr)
    {
        return GetOrCreateCache(ptr, ptr => new ElementDocument(ptr, false));
    }

    #endregion
}

File: RmlUi.Native/src/ElementDocument.cpp
#include "RmlNative.h"
#include "RmlUi/Core/ElementDocument.h"

RMLUI_CAPI void rml_ElementDocument_Show(Rml::ElementDocument* document, Rml::ModalFlag modal_flag = Rml::ModalFlag::None, Rml::FocusFlag focus_flag = Rml::FocusFlag::Auto) {
    document->Show(modal_flag, focus_flag);
}

RMLUI_CAPI void rml_ElementDocument_Hide(Rml::ElementDocument* document) {
    document->Hide();
}

RMLUI_CAPI void rml_ElementDocument_Close(Rml::ElementDocument* document) {
    document->Close();
}